sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/cardinality/ApproximateDistinct.java

```diff
@@ -0,0 +1,364 @@
+   * The longer the run is, the more unlikely it was to happen. Thus, the greater the number
+   * of elements that have been hashed before getting such a run.
+   * <br>Because this algorithm relies mainly on randomness, the stream is divided into buckets
+   * in order to reduce the variance of the estimation.
+   * Therefore, an estimation is applied on several samples and the overall estimation
+   * is then computed using an average (Harmonic mean).
+   *
+   * @param <InputT>    the type of the elements being combined
+   */
+  public static class ApproximateDistinctFn<InputT>
+      extends Combine.CombineFn<InputT, HyperLogLogPlus, HyperLogLogPlus> {
+
+    private final int p;
+
+    private final int sp;
+
+    private ApproximateDistinctFn(int p, int sp) {
+      this.p = p;
+      this.sp = sp;
+    }
+
+    /**
+     * Returns an {@code ApproximateDistinctFn} combiner with the given precision value p.
+     * This means that the input elements will be dispatched into 2^p buckets
+     * in order to estimate the cardinality.
+     *
+     * @param p           precision value for the normal representation
+     * @param <InputT>    the type of the input {@code Pcollection}'s elements being combined.
+     */
+    public static <InputT> ApproximateDistinctFn<InputT> create(int p) {
+      if (p < 4) {
+        throw new IllegalArgumentException("p must be greater than 4");
+      }
+      return new ApproximateDistinctFn<>(p, 0);
+    }
+
+    /**
+     * Returns an {@code ApproximateDistinctFn} combiner with the precision value p
+     * of this combiner and the given precision value sp for the sparse representation,
+     * meaning that the combiner will be using a sparse representation at small cardinalities.
+     *
+     * <p>The sparse representation does not initialize every buckets to 0 at the beginning.
+     * Indeed, for small cardinalities a lot of them will remain empty.
+     * Instead it builds a linked-list that grows when new indexes appear in the hashed values.
+     * To reduce collision of indexes and thus improve precision, we can define {@code sp > p}
+     * When the sparse representation would require more memory than the normal one,
+     * it is converted and the normal algorithm applies for the remaining elements.
+     *
+     * <p><b>WARNING : </b>Choose sp such that {@code p <= sp <= 32}
+     *
+     * <p>Example of use :
+     * <pre>{@code PCollection<Integer> input = ...;
+     * PCollection<HyperLogLogPlus> hllSketch = input
+     *        .apply(Combine.globally(ApproximateDistinct.ApproximateDistinctFn.<Integer>create(p)
+     *                .withSparseRepresentation(sp)));
+     * }</pre>
+     *
+     * @param sp          the precision of HyperLogLog+' sparse representation
+     */
+    public ApproximateDistinctFn<InputT> withSparseRepresentation(int sp) {
+      if (sp < p || sp > 32) {
+          throw new IllegalArgumentException("sp should be greater than p and lower than 32");
```

Comment:
<!--new_thread; commit:50a69e4b9213b300d0e3830e7c663d919d55c3ae; resolved:0-->
It is useful to spit out the arguments that the user gave us that are bad like:
```IllegalArgumentException("sp should be greater than p and lower than 32"); --> IllegalArgumentException(String.format("p <= sp <= 32, p = %s, sp = %s", p, sp));```

Also, the comment from the javadoc explains the restriction more clearly.